src/request_queue.js

```diff
@@ -240,165 +268,225 @@ export class RequestQueue {
 
-                return queueOperationInfo;
-            });
+        queueOperationInfo.request = request;
+
+        return queueOperationInfo;
     }
 
     /**
-     * Reclaims failed request back to the queue, so that it can be processed later again.
+     * Reclaims a failed request back to the queue, so that it can be returned for processed later again
+     * by another call to {@link RequestQueue#fetchNextRequest}.
      * The request record in the queue is updated using the provided `request` parameter.
-     * For example, this lets you store the number of retries for the request.
+     * For example, this lets you store the number of retries or error messages for the request.
      *
      * @param {Request} request
      * @param {Object} [options]
      * @param {Boolean} [options.forefront=false]
-     *   If `true` then requests get returned to the start of the queue
-     *   and to the back of the queue otherwise.
+     * If `true` then the request it placed to the beginning of the queue, so that it's returned
+     * in the next call to {@link RequestQueue#fetchNextRequest}. By default, it's put to the end of the queue.
      * @return {Promise<QueueOperationInfo>}
      */
-    reclaimRequest(request, options = {}) {
+    async reclaimRequest(request, options = {}) {
+        // TODO: This function should also support object instead of Apify.Request()
         const { forefront } = validateReclaimRequestParams(request, options);
 
-        // TODO: This function should also support object instead of Apify.Request()
+        if (!this.inProgress[request.id]) {
+            throw new Error(`Cannot reclaim request ${request.id}, because it is not in progress!`);
+        }
+
         // TODO: If request hasn't been changed since the last getRequest(),
         // we don't need to call updateRequest() and thus improve performance.
-        return requestQueues
-            .updateRequest({
-                request,
-                queueId: this.queueId,
-                forefront,
-                clientKey: this.clientKey,
-            })
-            .then((queueOperationInfo) => {
-                this._removeFromInProgress(request.id);
-                this._cacheRequest(getRequestId(request.uniqueKey), queueOperationInfo);
 
-                if (forefront) this.queueHeadDict.add(request.id, request.id, true);
+        const queueOperationInfo = await requestQueues.updateRequest({
+            request,
+            queueId: this.queueId,
+            forefront,
+            clientKey: this.clientKey,
+        });
 
-                queueOperationInfo.request = request;
+        this._cacheRequest(getRequestId(request.uniqueKey), queueOperationInfo);
+        queueOperationInfo.request = request;
 
-                return queueOperationInfo;
-            });
+        // Wait a little to increase a chance that the next call to fetchNextRequest() will return the request with updated data.
+        setTimeout(() => {
```

Comment:
Another random timeout. We need to figure out a way to make this clear and understandable.